lewiskyron/Yank
# ENG-TASK-63: Email & password authentication

## Description

This is quite a large PR of 81 files however a large percentage of the files are not relevant to the work that was actually done. Most of the files here are introduced due to the following changes:
- I have globalized the linting and formatting process
-  Additionally I have move the API connection to the LLM to a global introducing changes to files in both the extension and web app files. 
Also I am still working out which bits of the template I want to keep and which ones I want to get rid of. Hopefully as I further define which bits are useful for my use case the commits will be much smaller

That being said authenticating using supabase has not been easy. The learning experience however has been great and I am confident that this how now been well done. I have introduced a `middleware.ts` file which handles all the routing and api requests ensuring that users only access what is necessary. One thing to note however is that when attempting to fetch user data we cannot use the client based supasebase client. My assumption for now is that the supabase `serverClient` has some cookie handling making it authorized to request user info. To get around this I have used the `serverClient` to fetch the user object and propped it down to the components where we need to change or fetch user specific data.

## Type of Change


## Testing Instructions
Testing this can be a bit tricky. To login with email and password the auth flow only allows users to login into the page after  verifying their email. For now supabase has restricted the number of requests to 2 per min. Additionally emails for verification can only be sent team member emails. This is a bit annoying however I shall look into setting up an SMTP or having a dedicated email to do that. For now the signUp can only be tested by limited individuals. 
## Screenshots (if applicable)
<img width="1055" alt="Screenshot 2024-10-30 at 3 57 24 PM" src="https://github.com/user-attachments/assets/f55f7b94-146b-4a8d-8f42-5e0600f05ddb">
<img width="1171" alt="Screenshot 2024-10-30 at 3 57 01 PM" src="https://github.com/user-attachments/assets/aa89fcac-1d87-47df-8cd0-0a3095718cf1">
